File: applications/GentleApp/GentleApp/RemotePlatform.swift
/*
 * Copyright (c) 2025 GentleCare Corporation. All rights reserved.
 *
 * This source code and the accompanying materials are the confidential and
 * proprietary information of GentleCare Corporation. Unauthorized copying or
 * distribution of this file, via any medium, is strictly prohibited without
 * the prior written permission of John Payne.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import CoreBluetooth

enum PlatformMode: Float {
    case idle = 0.0
    case eversion = 1.0
    case eversionPaused = 2.0
    case inversion = 3.0
    case inversionPaused = 4.0
}

class RemotePlatform: NSObject, ObservableObject, CBCentralManagerDelegate, CBPeripheralDelegate {
    @Published var bluetoothStatus: String = "Starting..."
    @Published var isConnected: Bool = false
    @Published var connectedDeviceName: String = ""
    @Published var availableDevices: [String] = []

    @Published var airPressure: Float?
    @Published var pressureSensorError: Float?
    @Published var voltage: Float?
    @Published var motorPosition: Float?
    @Published var motorVelocity: Float?
    @Published var motorTorque: Float?
    @Published var motorError: Float?
    @Published var servoChamber: Float?
    @Published var mode: PlatformMode?
    @Published var progress: Float?
    
    private var availablePeripherals: [String:CBPeripheral] = [:]
    private var pollTimer: Timer!
    private var central: CBCentralManager!
    private var peripheral: CBPeripheral?
    private var characteristics: [CBUUID: CBCharacteristic] = [:]
    private var values: [CBUUID: Float] = [:]
    private var dataLogger: DataLogger!
    
    override init() {
        super.init()
        central = CBCentralManager(delegate: self, queue: nil)
        pollTimer = Timer.scheduledTimer(withTimeInterval: 0.1, repeats: true, block: { [weak self] _ in self?.poll() })
        dataLogger = try! DataLogger(columns: CHARACTERISTIC_UUIDS.map { $0.uuidString })
    }
    
    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        switch central.state {
        case .poweredOn:
            bluetoothStatus = "Searching for device..."
            startScan()
        case .poweredOff:
            bluetoothStatus = "Bluetooth is off. Please enable it in Settings."
        case .unauthorized:
            bluetoothStatus = "No Bluetooth permission. Grant it in Settings."
        case .unsupported:
            bluetoothStatus = "This device doesn't support Bluetooth."
        case .resetting:
            bluetoothStatus = "Bluetooth resetting..."
        default:
            bluetoothStatus = "Unknown error."
        }
    }
    
    private func startScan() {
        if central.state != .poweredOn {
            return;
        }
        
        availableDevices = []
        availablePeripherals = [:]
        central.scanForPeripherals(withServices: [SERVICE_UUID], options: [CBCentralManagerScanOptionAllowDuplicatesKey: true])
    }
    
    func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String : Any], rssi RSSI: NSNumber) {
        if peripheral.name == nil {
            return
        }
        
        availablePeripherals[peripheral.name!] = peripheral
        availableDevices = availablePeripherals.keys.sorted()
    }
    
    func connectToDevice(_ deviceName: String) {
        if availablePeripherals[deviceName] == nil {
            return
        }
        
        bluetoothStatus = "Connecting..."
        central.stopScan()
        peripheral = availablePeripherals[deviceName]!
        peripheral!.delegate = self
        central.connect(availablePeripherals[deviceName]!)
    }
    
    func disconnectFromDevice() {
        if !isConnected {
            return
        }
        
        bluetoothStatus = "Disconnected from device. Searching..."
        isConnected = false
        central.cancelPeripheralConnection(peripheral!)
        startScan()
    }
    
    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        bluetoothStatus = "Connected to device. Discovering services..."
        peripheral.discoverServices([SERVICE_UUID])
    }
    
    func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: (any Error)?) {
        bluetoothStatus = "Failed to connect. Retrying..."
        startScan()
    }
    
    func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: (any Error)?) {
        bluetoothStatus = "Disconnected from device. Searching..."
        isConnected = false
        startScan()
    }
    
    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: (any Error)?) {
        if error != nil || peripheral.services == nil {
            bluetoothStatus = "Service discovery error."
            return
        }
        
        bluetoothStatus = "Discovered services. Discovering characteristics..."
        peripheral.discoverCharacteristics(CHARACTERISTIC_UUIDS, for: peripheral.services![0])
    }
    
    func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: (any Error)?) {
        if error != nil || peripheral.services![0].characteristics == nil {
            bluetoothStatus = "Characteristic discovery error."
            return
        }
        
        for characteristic in peripheral.services![0].characteristics! {
            characteristics[characteristic.uuid] = characteristic
            if characteristic.properties.contains(.notify) {
                peripheral.setNotifyValue(true, for: characteristic)
            }
        }
        
        bluetoothStatus = "Connected to device."
        isConnected = true
        connectedDeviceName = peripheral.name!
    }
    
    func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: (any Error)?) {
        if error != nil || characteristic.value == nil {
            bluetoothStatus = "Failed to read value."
            return
        }
        
        if characteristic.value!.count != 4 {
            return
        }
        
        let value = Float(bitPattern: UInt32(littleEndian: characteristic.value!.withUnsafeBytes { $0.load(as: UInt32.self) }))
        values[characteristic.uuid] = value
    }
    
    private func poll() {
        if !isConnected {
            return
        }
        
        try! dataLogger.logRow(CHARACTERISTIC_UUIDS.map { values[$0] })
        
        airPressure = values[PRESSURE_SENSOR_UUID]
        pressureSensorError = values[PRESSURE_SENSOR_ERROR_UUID]
        voltage = values[DIMMER_VOLTAGE_UUID]
        motorPosition = values[MOTOR_POSITION_UUID]
        motorVelocity = values[MOTOR_VELOCITY_UUID]
        motorTorque = values[MOTOR_TORQUE_UUID]
        motorError = values[MOTOR_ERROR_UUID]
        servoChamber = values[SERVO_CHAMBER_UUID]
        mode = values[AUTO_CONTROL_MODE_UUID] == nil ? nil : PlatformMode(rawValue: values[AUTO_CONTROL_MODE_UUID]!)
        progress = values[AUTO_CONTROL_PROGRESS_UUID]
    }
    
    private func writeValue(_ value: Float, for uuid: CBUUID) {
        if peripheral == nil || characteristics[uuid] == nil {
            return
        }
        
        var bitPattern = value.bitPattern.littleEndian
        let data = withUnsafeBytes(of: &bitPattern) { Data($0) }
        peripheral!.writeValue(data, for: characteristics[uuid]!, type: .withResponse)
    }
    
    public func setMode(_ mode: PlatformMode) {
        writeValue(mode.rawValue, for: AUTO_CONTROL_MODE_UUID)
    }
    
    public func setChamber(_ chamber: Int) {
        writeValue(Float(chamber), for: SERVO_CHAMBER_UUID)
    }
    
    public func setVoltage(_ voltage: Float) {
        writeValue(voltage, for: DIMMER_VOLTAGE_UUID)
    }
    
    public func setPressureSetpoint(_ pressure: Float) {
        writeValue(pressure, for: PRESSURE_CONTROLLER_UUID)
    }
    
    public func setMotorVelocity(_ velocity: Float) {
        writeValue(velocity, for: MOTOR_VELOCITY_UUID)
    }
}

File: firmware/remote/src/remote_platform.cpp
/*
 * Copyright (c) 2025 GentleCare Corporation. All rights reserved.
 *
 * This source code and the accompanying materials are the confidential and
 * proprietary information of GentleCare Corporation. Unauthorized copying or
 * distribution of this file, via any medium, is strictly prohibited without
 * the prior written permission of GentleCare Corporation.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include <Arduino.h>
#include "config.h"
#include "remote_platform.h"
#include "common/uuids.h"

RemotePlatform::RemotePlatform(Adafruit_SSD1306 *display)
{
    this->found_device = false;
    this->display = display;
}

void RemotePlatform::start()
{
    NimBLEDevice::init("GentleBombDetonator");
    NimBLEDevice::setPower(ESP_PWR_LVL_P9);
    this->scanner = NimBLEDevice::getScan();
    this->scanner->setScanCallbacks(this);
    this->scanner->setActiveScan(true);
    this->scanner->setMaxResults(0);

    this->client = NimBLEDevice::createClient();
}

void RemotePlatform::update()
{
    if (!this->client->isConnected()) {
        this->display->clearDisplay();
        this->display->setCursor(0, 0);
        this->display->printf("Not connected\nSearching...");
        this->display->display();
    }

    while (!this->client->isConnected()) {
        if (!this->scanner->isScanning())
            this->scanner->start(0);

        if (this->found_device) {
            this->found_device = false;
            this->scanner->stop();
            this->display->clearDisplay();
            this->display->setCursor(0, 0);
            if (this->device->getName() == "Glide")
                this->display->println("Connected to \nGentleGlide");
            else if (this->device->getName() == "Wedge")
                this->display->println("Connected to \nGentleWedge");

            this->display->display();
            
            this->client->connect(this->device);
            this->client->setConnectionParams(6, 12, 0, 100);
            this->service = this->client->getService(SERVICE_UUID);
            for (int i = 0; i < CHARACTERISTIC_UUID_COUNT; i++) {
                this->characteristics[i] = this->service->getCharacteristic(CHARACTERISTIC_UUIDS[i]);
                if (this->characteristics[i] != nullptr && this->characteristics[i]->canNotify())
                    this->characteristics[i]->subscribe(true, std::bind(&RemotePlatform::on_notification, this, std::placeholders::_1, std::placeholders::_2, std::placeholders::_3, std::placeholders::_4));
                this->values[i] = 0.0;
            }
        }
    }
}

void RemotePlatform::onResult(const NimBLEAdvertisedDevice *device)
{
#if DEBUG_MODE
    Serial.printf("Device: %s\n", device->toString().c_str());
#endif
    if (device->getServiceUUID().toString() != SERVICE_UUID)
        return;
    if (device->getName() != CONNECTION_NAME)
        return;
    this->device = (NimBLEAdvertisedDevice *)device;
    this->found_device = true;
}

void RemotePlatform::on_notification(NimBLERemoteCharacteristic *characteristic, uint8_t *data, size_t length, bool isNotify)
{
    if (length != sizeof(float))
        return;

    for (int i = 0; i < CHARACTERISTIC_UUID_COUNT; i++) {
        if (strcmp(CHARACTERISTIC_UUIDS[i], characteristic->getUUID().toString().c_str()) == 0)
            this->values[i] = *(float *)data;
    }
}

NimBLERemoteCharacteristic *RemotePlatform::get_characteristic(const char *uuid)
{
    for (int i = 0; i < CHARACTERISTIC_UUID_COUNT; i++) {
        if (strcmp(CHARACTERISTIC_UUIDS[i], uuid) == 0)
            return this->characteristics[i];
    }

    return nullptr;
}

float RemotePlatform::get(const char *uuid)
{
    for (int i = 0; i < CHARACTERISTIC_UUID_COUNT; i++) {
        if (strcmp(CHARACTERISTIC_UUIDS[i], uuid) == 0)
            return this->values[i];
    }

    return 0.0;
}

void RemotePlatform::set(const char *uuid, float value, bool with_response)
{
    if (!this->client->isConnected())
        return;
    
    NimBLERemoteCharacteristic *characteristic = this->get_characteristic(uuid);
    if (characteristic == nullptr)
        return;

    characteristic->writeValue((uint8_t *)&value, sizeof(value), with_response);
}